File: Part1._Enumeration_problem/Lesson2._Generation_method/Binary_Strings.pas
program Binary_Strings;
const
    InputFile = 'BSTR.INP';
    OutputFile = 'BSTR.OUT';
    max = 30;
var
    x: array[1..max] of Integer;
    n, i: Integer;
    f: Text;
begin
    Assign(f, InputFile); Reset(f);
    ReadLn(f, n);
    Close(f);
    Assign(f, OutputFile); Rewrite(f);
    FillChar(x, SizeOf(x), 0); {Cấu hình ban đầu x=00..0}
    repeat {Thuật toán sinh}
        for i := 1 to n do Write(f, x[i]); {In ra cấu hình hiện tại}
        WriteLn(f);
        i := n; {x[i] là phần tử cuối dãy, lùi dần i cho tới khi gặp số 0 hoặc khi i = 0 thì dừng}
        while (i > 0) and (x[i] = 1) do Dec(i);
        if i > 0 then {Chưa gặp phải cấu hình 11…1}
        begin
            x[i] := 1; {Thay x[i] bằng số 1}
            FillChar(x[i + 1], (n - i) * SizeOf(x[1]), 0); {Đặt x[i+1] = x[i+2] = … = x[n] := 0}
        end;
    until i = 0; {Đã hết cấu hình}
    Close(f);
end.

File: Part1._Enumeration_problem/Lesson2._Generation_method/Binary_Strings.c
#include <stdio.h>

int main()
{
    FILE *fi = fopen("BSTR.INP", "r");
    if (fi != NULL)
    {
        int n = 0;
        fscanf(fi, "%d", &n);
        fclose(fi);
        if (n > 0 && n <= 30)
        {
            FILE *fo = fopen("BSTR.OUT", "w");
            int i = 0;
            int bit[n + 1];

            // Khởi tạo cấu hình đầu tiên
            for (i = 1; i <= n; i++)
            {
                bit[i] = 0;
            }

            do
            {
                // Xuất cấu hình hiện tại ra file
                for (i = 1; i <= n; i++)
                {
                    fprintf(fo, "%d", bit[i]);
                }
                fprintf(fo, "\n");

                // Sinh cấu hình kế tiếp
                i = n;
                while (i > 0 && bit[i] == 1)
                {
                    i--;
                }
                if (i > 0)
                {
                    bit[i] = 1;
                    for (int j = i + 1; j <= n; j++)
                    {
                        bit[j] = 0;
                    }
                }

            } while (i != 0); // Gặp cấu hình cuối cùng

            fclose(fo);
        }
    }
    return 0;
}